cpp/src/arrow/types/struct.h

```diff
@@ -23,7 +23,122 @@
 #include <vector>
 
 #include "arrow/type.h"
+#include "arrow/types/list.h"
+#include "arrow/types/primitive.h"
 
-namespace arrow {}  // namespace arrow
+namespace arrow {
+
+class StructArray : public Array {
+ public:
+  StructArray(const TypePtr& type, int32_t length, std::vector<ArrayPtr>& field_arrays,
+      int32_t null_count = 0, std::shared_ptr<Buffer> null_bitmap = nullptr)
+      : Array(type, length, null_count, null_bitmap) {
+    type_ = type;
+    field_arrays_ = field_arrays;
+  }
+
+  Status Validate() const override;
+
+  virtual ~StructArray() {}
+
+  // Return a shared pointer in case the requestor desires to share ownership
+  // with this array.
+  const std::shared_ptr<Array>& field(int32_t pos) const { return field_arrays_.at(pos); }
+  const std::vector<ArrayPtr>& fields() const { return field_arrays_; }
+
+  const std::shared_ptr<DataType>& field_type(int32_t pos) const {
+    return field_arrays_.at(pos)->type();
+  }
+
+  bool EqualsExact(const StructArray& other) const;
+  bool Equals(const std::shared_ptr<Array>& arr) const override;
+
+ protected:
+  // The child arrays corresponding to each field of the struct data type.
+  std::vector<ArrayPtr> field_arrays_;
+};
+
+// ---------------------------------------------------------------------------------
+// StructArray builder
+class StructBuilder : public ArrayBuilder {
+ public:
+  StructBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
+      std::vector<std::shared_ptr<ArrayBuilder>>& field_builders)
+      : ArrayBuilder(pool, type) {
+    field_builders_ = field_builders;
+  }
+
+  Status Init(int32_t elements) override {
+    RETURN_NOT_OK(ArrayBuilder::Init(elements));
+    return Status::OK();
+  }
+
+  Status Resize(int32_t capacity) override {
+    // Need space for the end offset
+    if (capacity < MIN_BUILDER_CAPACITY) { capacity = MIN_BUILDER_CAPACITY; }
+
+    if (capacity_ == 0) {
+      RETURN_NOT_OK(ArrayBuilder::Init(capacity));
+    } else {
+      RETURN_NOT_OK(ArrayBuilder::Resize(capacity));
+    }
+    capacity_ = capacity;
+
+    for (auto it : field_builders_) {
+      RETURN_NOT_OK(it->Resize(capacity));
+    }
+    return Status::OK();
+  }
+
+  // null_bitmap is of equal length to every child field, and any zero byte
+  // will be considered as a null for that field, but users must using app-
+  // end methods or advance methods of the child builders independently to
+  // insert data.
+  Status Append(const uint8_t* null_bitmap, int32_t length) {
+    RETURN_NOT_OK(Reserve(length));
+    UnsafeAppendToBitmap(null_bitmap, length);
+    return Status::OK();
+  }
+
+  std::shared_ptr<Array> Finish() override {
+    DCHECK(field_builders_.size());
```

Comment:
Yep, it's sort of esoteric use case, but the code shouldn't break when all the fields are dropped as the result of some selection / filter operation (we permit data frames with length but no columns in Python for example)